pg_analytics/src/hooks/createfunction.rs

```diff
@@ -0,0 +1,220 @@
+            // Prefer to use std::mem::offset_of! but it is still unstable as of rust 1.76.0.
+            let fcinfo_size = max(
+                std::mem::size_of::<pg_sys::FunctionCallInfoBaseData>(),
+                memoffset::offset_of!(pg_sys::FunctionCallInfoBaseData, args)
+                    + std::mem::size_of::<pg_sys::NullableDatum>() * (num_args - 1),
+            );
+            let fcinfo = pg_sys::palloc0(fcinfo_size) as *mut pg_sys::FunctionCallInfoBaseData;
+            (*fcinfo).flinfo = flinfo.as_ptr();
+            (*fcinfo).context = std::ptr::null_mut();
+            (*fcinfo).resultinfo = std::ptr::null_mut();
+            (*fcinfo).fncollation = pg_sys::Oid::INVALID;
+            (*fcinfo).isnull = false;
+            (*fcinfo).nargs = num_args as i16 - 1;
+
+            // Call function on each set of arguments and push to result vector
+            let mut result_vec = vec![];
+            for row_index in 0..num_rows {
+                for arg_index in 1..num_args {
+                    let fcinfo_arg = (*fcinfo).args.as_mut_ptr().add(arg_index - 1);
+                    (*fcinfo_arg).value = arg_arrays[arg_index - 1]
+                        .get_datum(row_index)?
+                        .ok_or(DataFusionError::Internal("No datum for type".to_string()))?;
+                    (*fcinfo_arg).isnull = false;
+                }
+
+                let result_datum = (*(*fcinfo).flinfo)
+                    .fn_addr
+                    .ok_or(DataFusionError::Internal(
+                        "Invalid function address for udf".to_string(),
+                    ))?(fcinfo);
+                result_vec.push((result_datum, false));
+            }
+
+            // UDF return type
+            let rettype = pg_sys::get_func_rettype(func_oid);
+
+            Ok(ColumnarValue::Array(
+                result_vec
+                    .into_iter()
+                    .into_arrow_array(rettype.into(), PgTypeMod(-1))?,
+            ))
+        } else {
+            Err(DataFusionError::Internal("No funcname".to_string()))
+        }
+    });
+
+    drop(memory_context);
+
+    ret
+}
+
+pub unsafe fn loadfunction(funcname: &str) -> Result<(), ParadeError> {
+    // Register all overloads with this function name
+    let mut func_candidate = func_list_from_name(funcname)?;
+    while !func_candidate.is_null() {
+        let func_oid = (*func_candidate).oid;
+        let arg_types = (*func_candidate).args.as_mut_ptr();
+        let nargs = (*func_candidate).nargs;
+        let ret_oid = pg_sys::get_func_rettype(func_oid);
+
+        // Create vector of input types
+        let mut input_types = vec![];
+        input_types.push(DataType::Utf8); // function name
+        for param_index in 0..nargs {
+            let arg_oid = arg_types.add(param_index as usize);
+            let ArrowDataType(datatype) =
+                PgAttribute((*arg_oid).into(), PgTypeMod(-1)).try_into()?;
```

Comment:
Does hardcoding the typemod cause any edge cases?

Reply:
Nope - there's nothing that happens between conversion to and from Datafusion data type. I added a clarifying comment, though.